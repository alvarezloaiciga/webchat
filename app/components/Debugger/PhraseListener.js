// @flow

import React from 'react';

type listener = {
  phrase: string,
  callback: Function,
  caseInsensitive: boolean,
};

type PhraseListenerProps = {
  listeners: Array<listener>,
  timeout: number,
};

type PhraseListenerState = {
  keys: string,
};

export class PhraseListener extends React.Component {
  props: PhraseListenerProps;
  state: PhraseListenerState = {keys: ''};
  timeoutFunctionId: number;

  static defaultProps = {timeout: 5000};

  componentDidMount() {
    window.addEventListener('keydown', this.handleKeyDown);
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutFunctionId);
    window.removeEventListener('keydown', this.handleKeyDown);
  }

  resetTimeout = () => {
    clearTimeout(this.timeoutFunctionId);
    this.timeoutFunctionId = setTimeout(() => {
      this.setState({keys: ''});
    }, this.props.timeout);
  };

  handleKeyDown = (e: {keyCode: number}) => {
    this.resetTimeout();

    this.setState({keys: this.state.keys + String.fromCharCode(e.keyCode).toLowerCase()});
    this.props.listeners.forEach(l => {
      const {phrase, callback, caseInsensitive} = l;

      if (
        this.state.keys.includes(phrase) ||
        (caseInsensitive && this.state.keys.toLowerCase().includes(phrase.toLowerCase()))
      ) {
        callback();
        this.setState({keys: ''});
      }
    });
  };

  render() {
    return null;
  }
}

export default PhraseListener;
