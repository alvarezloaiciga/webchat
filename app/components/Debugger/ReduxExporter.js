// @flow

import React from 'react';
import {getFormattedDateAndTime} from 'utils/utils';
import {saveAs} from 'file-saver';

type ReduxExporterProps = {
  monitorState?: {initialScrollTop: number}, // eslint-disable-line react/no-unused-prop-types
  actionsById?: Object,
  nextActionId?: number,
  stagedActionIds?: Array<number>,
  skippedActionIds?: Array<number>,
  committedState?: Object,
  computedStates?: Array<Object>, // eslint-disable-line react/no-unused-prop-types
  currentStateIndex?: number,
};

export class ReduxExporter extends React.Component {
  props: ReduxExporterProps;

  exportReduxState = () => {
    saveAs(
      new Blob([JSON.stringify(this.getStateAndActions(), null, 2)], {
        type: 'application/json;charset=utf-8',
      }),
      `chat-${getFormattedDateAndTime(new Date())}`,
    );
  };

  getStateAndActions = () => ({
    monitorState: this.props.monitorState,
    actionsById: this.props.actionsById,
    nextActionId: this.props.nextActionId,
    stagedActionIds: this.props.stagedActionIds,
    skippedActionIds: this.props.skippedActionIds,
    committedState: this.props.committedState,
    computedStates: [],
    currentStateIndex: this.props.currentStateIndex,
  });

  render() {
    return (
      <div className="ReduxExporter">
        <i
          className={`fa fa-download icon`}
          title="Export Redux State"
          onClick={this.exportReduxState}
        />
      </div>
    );
  }
}

export default ReduxExporter;
