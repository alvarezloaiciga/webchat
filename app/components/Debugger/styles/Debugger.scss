.Debugger {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 5px 0;
  background-color: white;
  border-bottom: 1px solid #7e7e7e;
  font-size: 12px;

  .lhsIcons {
    display: flex;
    flex: 1 0 auto;
  }

  .rhsIcons {
    display: flex;
    align-items: center;

    .versions {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: flex-end;
      margin-right: 10px;
    }
  }

  i {
    color: lighten(black, 40%);
    margin: 0 5px;

    &:hover {
      color: black;
      cursor: pointer;
    }
  }
}
