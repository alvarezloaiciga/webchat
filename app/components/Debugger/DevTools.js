import React from 'react';
import LogMonitor from 'redux-devtools-log-monitor';
import {createDevTools} from 'redux-devtools';
import ReduxExporter from './ReduxExporter';

export const MonitorTools = createDevTools(<LogMonitor />);
export default createDevTools(<ReduxExporter />);
