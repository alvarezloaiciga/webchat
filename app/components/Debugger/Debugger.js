// @flow
declare var __VERSION__: string;
import React from 'react';
import QUIQ from 'utils/quiq';
import {inNonProductionCluster, inLocalDevelopment} from 'utils/utils';
import DevTools from './DevTools';
import PhraseListener from './PhraseListener';
import {getChatClient} from '../../ChatClient';
import {version} from '../../../node_modules/quiq-chat/package.json';
import './styles/Debugger.scss';
import type {QuiqChatClientType} from 'quiq-chat';

type DebuggerState = {
  hidden: boolean,
};

export class Debugger extends React.Component {
  state: DebuggerState = {
    hidden: !QUIQ.DEBUG,
  };
  client: QuiqChatClientType;

  constructor() {
    super();
    this.client = getChatClient();
  }

  shouldShowDebugger = () => inNonProductionCluster() || inLocalDevelopment();

  renderPhraseListener = () =>
    <PhraseListener
      listeners={[
        {
          phrase: 'quiqdebug',
          callback: () => this.setState({hidden: !this.state.hidden}),
          caseInsensitive: true,
        },
      ]}
    />;

  render() {
    if (!this.shouldShowDebugger()) return null;
    if (this.state.hidden) return this.renderPhraseListener();

    return (
      <div className="Debugger">
        {this.renderPhraseListener()}
        <div className="lhsIcons">
          <DevTools />
          {
            <i
              className={`fa fa-sign-in icon`}
              title="AndrewTest Auth User via Secure Cookie (Old Deprecated Way)"
              onClick={this.client.DEPRECATED_AUTH_USER_DO_NOT_USE}
            />
          }
        </div>
        <div className="rhsIcons">
          <div className="versions">
            <span>
              WC: v{__VERSION__}
            </span>
            <span>
              QC: v{version}
            </span>
          </div>
          <i
            className={`fa fa-close icon`}
            title="Close Debugger"
            onClick={() => this.setState({hidden: true})}
          />
        </div>
      </div>
    );
  }
}

export default Debugger;
