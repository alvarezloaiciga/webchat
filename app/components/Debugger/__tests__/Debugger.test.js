// @flow
jest.mock('../../../ChatClient');
import React from 'react';
import Debugger from '../Debugger';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';

describe('Debugger component', () => {
  let wrapper: ShallowWrapper;
  let render: () => void;

  beforeEach(() => {
    render = () => {
      wrapper = shallow(<Debugger />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
