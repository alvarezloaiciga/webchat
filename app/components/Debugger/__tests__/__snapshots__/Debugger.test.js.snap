// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debugger component rendering renders phrase listener 1`] = `
<PhraseListener
  listeners={
    Array [
      Object {
        "callback": [Function],
        "caseInsensitive": true,
        "phrase": "quiqdebug",
      },
    ]
  }
  timeout={5000}
/>
`;

exports[`Debugger component rendering when not hidden renders Debugger bar 1`] = `
<div
  className="Debugger"
>
  <PhraseListener
    listeners={
      Array [
        Object {
          "callback": [Function],
          "caseInsensitive": true,
          "phrase": "quiqdebug",
        },
      ]
    }
    timeout={5000}
  />
  <div
    className="lhsIcons"
  >
    <DevTools />
    <i
      className="fa fa-sign-in icon"
      onClick={[Function]}
      title="AndrewTest Auth User via Secure Cookie (Old Deprecated Way)"
    />
  </div>
  <div
    className="rhsIcons"
  >
    <div
      className="versions"
    >
      <span>
        WC: v
        1.0.0
      </span>
      <span>
        QC: v
        1.11.3
      </span>
    </div>
    <i
      className="fa fa-close icon"
      onClick={[Function]}
      title="Close Debugger"
    />
  </div>
</div>
`;
